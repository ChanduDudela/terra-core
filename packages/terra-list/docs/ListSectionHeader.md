# List Section Header

The list section header is a sub-component of the list section. It provides the visual representation of the section, but does not provided any collapsible functionality itself, it's collapsible props are for the display of the collapse/expand indicator.

When consumed by itself as an individual list item avoid enabling the collapsible properties.

## Usage

```jsx
import React from 'react';
import List, { SectionHeader } from 'terra-list/lib/List';

<SectionHeader
  title="test title"
  key="section-1"
/>

```
