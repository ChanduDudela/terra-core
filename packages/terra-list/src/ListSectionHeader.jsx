import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';
import ListUtils from './ListUtils';
import styles from './ListSectionHeader.module.scss';

const cx = classNames.bind(styles);

/* eslint-disable react/forbid-prop-types */
const propTypes = {
  /**
   * Whether or not the section is collapsed.
   */
  isCollapsed: PropTypes.bool,
  /**
   * Whether or not the section can be collapsed.
   */
  isCollapsible: PropTypes.bool,
  /**
   * Optionally sets the heading level. One of `1`, `2`, `3`, `4`, `5`, `6`. Default `level=2`.
   */
  level: PropTypes.oneOf([1, 2, 3, 4, 5, 6]),
  /**
   * The associated metaData to be provided in the onSelect callback.
   */
  metaData: PropTypes.object,
  /**
   * Function callback for when the appropriate click or key action is performed.
   * Callback contains the javascript evnt and prop metadata, e.g. onSelect(event, metaData)
   */
  onSelect: PropTypes.func,
  /**
   * Function callback passthrough for the ref of the section li.
   */
  refCallback: PropTypes.func,
  /**
   * Header to be placed within a section list item
   */
  title: PropTypes.string,
};
/* eslint-enable react/forbid-prop-types */

const defaultProps = {
  isCollapsed: false,
  isCollapsible: false,
  level: 2,
  title: '',
};

const SectionHeader = ({
  isCollapsed,
  isCollapsible,
  level,
  metaData,
  onBlur,
  onClick,
  onKeyDown,
  onMouseDown,
  onSelect,
  refCallback,
  title,
  ...customProps
}) => {
  const sectionHeaderClassNames = cx([
    'section-header',
    { 'is-collapsible': isCollapsible },
    customProps.className,
  ]);

  const ariaSpread = {};
  const Element = `h${level}`;
  let titleElement = <Element className={cx('title')}>{title}</Element>;
  let accordionIcon;
  if (isCollapsible) {
    accordionIcon = (
      <div className={cx('start')}>
        <span className={cx(['accordion-icon', { 'is-open': !isCollapsed }])} />
      </div>
    );
    titleElement = (
      <div className={cx('fill')}>
        {titleElement}
      </div>
    );

    ariaSpread.onClick = ListUtils.wrappedOnClickForItem(onClick, onSelect, metaData);
    ariaSpread.onKeyDown = ListUtils.wrappedOnKeyDownForItem(onKeyDown, onSelect, metaData);
    ariaSpread.tabIndex = '0';
    ariaSpread.role = 'heading';
    ariaSpread['aria-expanded'] = !isCollapsed;
    ariaSpread['aria-level'] = 1;
    ariaSpread.onBlur = ListUtils.wrappedOnBlur(onBlur, event => event.currentTarget.setAttribute('data-focusable', 'true'));
    ariaSpread.onMouseDown = ListUtils.wrappedOnMouseDown(onMouseDown, event => event.currentTarget.setAttribute('data-focusable', 'false'));
  }

  return (
    <li {...customProps} {...ariaSpread} className={sectionHeaderClassNames} ref={refCallback}>
      {accordionIcon}
      {titleElement}
    </li>
  );
};

SectionHeader.propTypes = propTypes;
SectionHeader.defaultProps = defaultProps;

export default SectionHeader;
