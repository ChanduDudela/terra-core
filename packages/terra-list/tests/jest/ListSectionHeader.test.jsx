import React from 'react';
import ListSectionHeader from '../../src/ListSectionHeader';

// Snapshot Tests
it('should render default', () => {
  const shallowComponent = shallow(<ListSectionHeader />);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render with title', () => {
  const shallowComponent = shallow(<ListSectionHeader title="Test Title" />);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render with title and level', () => {
  const shallowComponent = shallow(<ListSectionHeader title="Test Title" level={3} />);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render with isCollapsed', () => {
  const shallowComponent = shallow(<ListSectionHeader isCollapsed />);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render with isCollapsible', () => {
  const shallowComponent = shallow(<ListSectionHeader isCollapsible />);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render with callback functions', () => {
  const shallowComponent = shallow(<ListSectionHeader refCallback={jest.fn()} onSelect={jest.fn()} />);
  expect(shallowComponent).toMatchSnapshot();
});
