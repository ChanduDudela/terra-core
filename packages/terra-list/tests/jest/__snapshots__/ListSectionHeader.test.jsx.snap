// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render default 1`] = `
<li
  className="section-header"
>
  <h1
    className="title"
  />
</li>
`;

exports[`should render with callback functions 1`] = `
<li
  className="section-header"
>
  <h1
    className="title"
  />
</li>
`;

exports[`should render with isCollapsed 1`] = `
<li
  className="section-header"
>
  <h1
    className="title"
  />
</li>
`;

exports[`should render with isCollapsible 1`] = `
<li
  aria-expanded={true}
  aria-level={1}
  className="section-header is-collapsible"
  data-focusable="true"
  onBlur={[Function]}
  onMouseDown={[Function]}
  role="heading"
  tabIndex="0"
>
  <div
    className="start"
  >
    <span
      className="accordion-icon is-open"
    />
  </div>
  <div
    className="fill"
  >
    <h1
      className="title"
    />
  </div>
</li>
`;

exports[`should render with title 1`] = `
<li
  className="section-header"
>
  <h1
    className="title"
  >
    Test Title
  </h1>
</li>
`;

exports[`should render with title and level 1`] = `
<li
  className="section-header"
>
  <h3
    className="title"
  >
    Test Title
  </h3>
</li>
`;
